### ISSUE TITLE
Error in matrix dimension mismatch

### ISSUE THREAD

**Issue by xboldocky** [pnnl/neuromancer#14]
Hi @drgona,

I'm trying to design a DPC controller. However, I'm facing a runtime error message that indicates a mismatch in matrix dimensions. I'm trying to resolve it, unfortunately with no success so far. It would be very helpful if you could take a look at it. Here is a python file: [magnet_DPC_LSSM_var_ref.zip](https://github.com/pnnl/neuromancer/files/10824497/magnet_DPC_LSSM_var_ref.zip).

Error message:
```diff
Traceback (most recent call last):
  File "neuromancer/examples/control/magnetoshield_DPC_LSSM_var_ref.py", line 351, in <module>
    trajectories = cl_sim.simulate(nsim=sim_steps, data_init=data_init, data_traj=data_traj)
  File "neuromancer/neuromancer/simulator.py", line 103, in simulate
    data_step = self.step(data)
  File "neuromancer/neuromancer/simulator.py", line 111, in step
    output_dict = component.step(input_dict)
  File "neuromancer/neuromancer/simulator.py", line 56, in step
    data = self._forward_step(data)
  File "neuromancer/neuromancer/simulator.py", line 588, in _forward_step
    u = self.eval_policy(features)
  File "neuromancer/neuromancer/simulator.py", line 652, in eval_policy
    u = self.policy(xi)
  File "anaconda3/envs/neuromancer/lib/python3.10/site-packages/torch/nn/modules/module.py", line 1130, in _call_impl
    return forward_call(*input, **kwargs)
  File "neuromancer/neuromancer/blocks.py", line 133, in forward
    x = nlin(lin(x))
  File "anaconda3/envs/neuromancer/lib/python3.10/site-packages/torch/nn/modules/module.py", line 1130, in _call_impl
    return forward_call(*input, **kwargs)
  File "slim/slim/linear.py", line 115, in forward
    return self.linear(x)
  File "anaconda3/envs/neuromancer/lib/python3.10/site-packages/torch/nn/modules/module.py", line 1130, in _call_impl
    return forward_call(*input, **kwargs)
  File "anaconda3/envs/neuromancer/lib/python3.10/site-packages/torch/nn/modules/linear.py", line 114, in forward
    return F.linear(input, self.weight, self.bias)
RuntimeError: mat1 and mat2 shapes cannot be multiplied (1x4 and 13x20)

**Comment by aarontuor** [comment]
We are currently undergoing a refactor in the closed loop system modeling to make implementation of controllers much simpler for the end user. We will close this issue with comment once the new control interface is completed in the next couple weeks. We would be happy to work with you to help implement your controller once the the new interface is complete.

**Comment by drgona** [comment]
The underlying problem is with the current implementation of the data processing method in the SystemSimulator class.
The current implementation select_data_step method selects a single step from a given input trajectory, hence does not support handling of n-step long trajectories with n>1 
https://github.com/pnnl/neuromancer/blob/master/neuromancer/simulator.py#L106

This issue will be fixed in the upcoming release.

**Comment by camilogonzalez97** [comment]
Hi @drgona ,

I recently encountered this error with v1.4.2 while trying to implement DPC. I have been using a plant model consisting of a NODE for state transition and a multi layer perceptron for the output equation. When I try to train the policy as part of the closed loop system the train function errors out with either: the matrix multiplication error referenced in this issue, or with an "out of bounds for dimensions 1 with size 1". The latter is raised by [this line](https://github.com/pnnl/neuromancer/blob/e0f02e592ce32584a03fba52d422aad4bbb4276d/src/neuromancer/system.py#L254).

Could you explain what the expected tensor dimensions are for training policies in closed loop when controlling outputs and not states. I am currently using an initial measured output and references for each output over the horizon. During training the policy only outputs 1 set of inputs and the plant model errors out because it expects inputs over the entire horizon. I haven't been able to make the policy generate inputs over the entire horizon. Any help would be much appreciated.

**Comment by drgona** [comment]
Hi, @camilogonzalez97, for training open-loop system ID and closed-loop control policies using the [System class](https://github.com/pnnl/neuromancer/blob/e0f02e592ce32584a03fba52d422aad4bbb4276d/src/neuromancer/system.py#L95) we expect 3D tensors, where the first dimension is batch number, the second is prediction horizon (rollout) length, and the third is the dimension of the variable.

I suggest you have a closer look at our [control examples](https://github.com/pnnl/neuromancer/tree/e0f02e592ce32584a03fba52d422aad4bbb4276d?tab=readme-ov-file#control)
If the problem prevails please create new issue with minimum example of the code that runs into this issue.
